### ISSUE TITLE
Respect freedesktop dark style in v4.8.3

### ISSUE THREAD

**Issue by TomiOhl** [telegramdesktop/tdesktop#26370]
### Steps to reproduce

1. Make sure "Match the system settings" is checked in Chat settings
2. Toggle the theme on your system

### Expected behaviour

Telegram should follow the dark/light style

### Actual behaviour

Stays always light. Worked fine in 4.8.2, this only happens since 4.8.3.

### Operating system

Elementary OS 7, Pantheon

### Version of Telegram Desktop

4.8.3

### Installation source

Static binary from official website

**Comment by ilya-fedin** [comment]
Works just fine for me on KDE

**Comment by TomiOhl** [comment]
What I've tried: deleted Telegram along with the files it created in $HOME/.local/share. Reinstalled from the binary on the official website. The problem still persists.
The logs when I run the executable:
```sh
tomiohl@ThinkPad-T16:~$ /opt/Telegram/Telegram
QApplication: invalid style override 'adwaita' passed, ignoring it.
	Available styles: Windows, Fusion
QPainter::begin: Paint device returned engine == 0, type: 2
QWidget::render: Cannot render with an inactive painter
qt.gui.imageio.jpeg: Corrupt JPEG data: premature end of data segment

(Telegram:8668): Gtk-WARNING **: 23:42:13.110: Theme parsing error: <data>:1:0: Expected a valid selector
```

**Comment by ilya-fedin** [comment]
I still can't reproduce. I change between light and dark color scheme in KDE settings and tdesktop reacts as it should.

**Comment by TomiOhl** [comment]
Hmm, well, for me all apps work (including Chrome since 114), but tdesktop doesn't for some reason. Do you have similar logs? I suspect these might be related.

**Comment by ilya-fedin** [comment]
Not all of those lines but there's nothing bad in those lines.

**Comment by ilya-fedin** [comment]
> What I've tried: deleted Telegram along with the files it created in $HOME/.local/share.

I hope you know this feature is opt-in

**Comment by TomiOhl** [comment]
> I hope you know this feature is opt-in
If you mean that it needs to be enabled in the settings, refer to the first point in the issue:
> Make sure "Match the system settings" is checked in Chat settings
And yes, I've re-enabled it after deleting those files.

> I change between light and dark color scheme in KDE settings and tdesktop reacts as it should.
I've just discovered that it does react for me too when the app is opened and I change the settings. So, let's some it up with the new findings:
- when the app starts, it's always light (no matter the system settings + relevant option enabled in tdesktop)
- if I toggle the "Match the system settings" option, it stays light even if the system is in dark mode
- if I change between dark/light mode in settings while the app is running, it reacts as it should
- if I close the app and reopen it, it will be light again even if the system is in dark mode

**Comment by ilya-fedin** [comment]
If I switch the system to dark color scheme, it starts with night mode for me.

**Comment by ilya-fedin** [comment]
Maybe elementaryOS has a broken spec implementation?

**Comment by TomiOhl** [comment]
Seems to be something DE-specific if it really works on KDE.
However, as I've said this worked perfectly fine with the previous version of tdesktop (and all of them since it was added to the app) and it works perfectly fine with every other app I'm using.

**Comment by Aokromes** [comment]
> Reference in

try downloading portable previous version of telegram desktop to see if the bug exist there. it's not the 1st time one system update breaks things.

**Comment by TomiOhl** [comment]
Thanks for the tip, I've downloaded [v4.8.1](https://github.com/telegramdesktop/tdesktop/releases/tag/v4.8.1) and it works perfectly fine. So I can confirm this is not an issue with the system.

**Comment by ilya-fedin** [comment]
It could be still an issue with the system. Maybe your DE implements the spec in a wrong way.

**Comment by TomiOhl** [comment]
Well, the elementary OS devs were the ones who proposed the dark style preference to freedesktop, but ofc that doesn't mean that they don't do something differently than others.
I understand tdesktop works perfectly on KDE and Gnome, too (just tried on Ubuntu).
At the same time, literally all apps work perfectly on elementary OS's Pantheon, including tdesktop up until v4.8.1.
This implies there was some (linux-specific) update to tdesktop that broke it.

**Comment by ilya-fedin** [comment]
Or to some of its dependency.

**Comment by q234rty** [comment]
Having a similar issue:

I'm using [darkman] as my portal implementation. If I run `darkman toggle` when tdesktop is running, it reacts just fine. However if I do so when tdesktop is not running, tdesktop seems to ignore that and continuing using the old color scheme when it starts.

**Comment by ilya-fedin** [comment]
@q234rty are you using an official build?

**Comment by q234rty** [comment]
No, and it doesn't seem to be reproducible on an official build. If I use `QT_QPA_PLATFORMTHEME=xdgdesktopportal` on an arch build it also works fine, so I guess this is related to https://github.com/desktop-app/patches/blob/master/qtbase_6.5.1/0015-nicer-platformtheme-choosing.patch ?

**Comment by ilya-fedin** [comment]
Yeah, it's related.

**Comment by ilya-fedin** [comment]
You apparently have to bug Qt (https://bugreports.qt.io) if you believe xdgdesktopportal is the right QPA for your environment so they make it default for you updstream

**Comment by TomiOhl** [comment]
Nice, seems like this is the problem on elementary, too. Gonna make a bug report for qt. Thanks

**Comment by ilya-fedin** [comment]
It's not if you're using the official binary as you've specified, it has a patch as was mentioned earlier.

**Comment by TomiOhl** [comment]
Yeah I forgot about that detail. Went ahead and tried out the latest build (4..8.4) with QT_QPA_PLATFORMTHEME=xdgdesktopportal and system set to dark mode, tdesktop started in dark mode.
As a side effect the match system settings toggle disappeared until I've launched tdesktop without the env variable. I definitely don't fully understand the situation but this certainly had some effect.
Another side effect when I've launched it with the platformtheme set was that it didn't output that gtk warning with the theme parsing error to the terminal.

**Comment by ilya-fedin** [comment]
hm, do you have QT_QPA_PLATFORMTHEME pre-defined by your OS to any value? It would of course override the patched logic.

**Comment by TomiOhl** [comment]
The default value is `gtk3`

**Comment by ilya-fedin** [comment]
Apparently you have to report that to your distro then? Although I'm not sure that `xdgdesktopportal` value would be a better one: Qt doesn't have Pantheon in its hardcoded list of gtk environments and so `xdgdesktopportal` won't fallback to `gtk3` for integration bits not covered by the portal. I'm not sure what to suggest, you better report that to your distro so they get in contact with Qt and help them develop a better UX for Pantheon users.

**Comment by ilya-fedin** [comment]
I don't know though whether your distro accepts reports for non-repo software, so that might be a quest that can end up in a requirement to use tdesktop provided by your distro that likely uses older Qt and doesn't have such issue yet (but it's likely to re-use the highly patched tdesktop version from Debian that removes some features like background image synchronization, tg:// handler registraion, noise suppresion in calls, it's also likely very old)

**Comment by ilya-fedin** [comment]
This is Qt's logic to determine the color scheme using gtk3:
https://github.com/qt/qtbase/blob/v6.5.1/src/plugins/platformthemes/gtk3/qgtk3storage.cpp#L341-L342
https://github.com/qt/qtbase/blob/v6.5.1/src/plugins/platformthemes/gtk3/qgtk3interface.cpp#L498-L512

This likely means that the foreground is lighter than the background for the GtkEntry widget.

`QT_LOGGING_RULES=qt.qpa.gtk.debug=true` should help to determine why exactly it has decided that your theme is dark.

**Comment by TomiOhl** [comment]
Ok so I've discovered one thing: if I set `gtk-application-prefer-dark-theme` to 1, tdesktop opens with dark theme. So apparently that's what the gtk3 platformtheme does

**Comment by ilya-fedin** [comment]
Yeah, it reads this setting but also reads your theme colors so it should work just fine without that. I've mentioned in the previous post the environment variable that should help to debug that.

**Comment by TomiOhl** [comment]
The output when tdesktop is launched with `QT_LOGGING_RULES=qt.qpa.gtk.debug=true`:
```
qt.qpa.gtk: Color map populated from defaults.
qt.qpa.gtk: GTK theme initialized: "io.elementary.stylesheet.blueberry" Qt::ColorScheme::Light
qt.qpa.gtk: Creating palette: "SystemPalette"
qt.qpa.gtk: System Palette defined "io.elementary.stylesheet.blueberry" Qt::ColorScheme::Light QPalette(resolve=0x74bfffa7fffd3fff,"WindowText:[Active:#ff333333,Disabled:#ff191919,Inactive:#ff333333],Button:[Active:#fff1f1f1,Disabled:#ffd4d0c8,Inactive:#ffd4d0c8],Light:[Active:#ffffffff,Disabled:#ffffffff,Inactive:#ffffffff],Midlight:[Active:#ffe9e7e3,Disabled:#ffe9e7e3,Inactive:#ffe9e7e3],Dark:[Active:#ffc1c1c1,Disabled:#ff6a6864,Inactive:#ff6a6864],Mid:[Active:#ffa0a0a4,Disabled:#ffa0a0a4,Inactive:#ffa0a0a4],Text:[Active:#ff333333,Disabled:#ff191919,Inactive:#ff333333],BrightText:[Active:#ffffffff,Disabled:#ffffffff,Inactive:#ffffffff],ButtonText:[Active:#ff333333,Disabled:#ff333333,Inactive:#ff191919],Base:[Active:#fff1f1f1,Disabled:#ffd4d0c8,Inactive:#fff1f1f1],Window:[Active:#fff1f1f1,Disabled:#ffd4d0c8,Inactive:#fff1f1f1],Shadow:[Active:#ffaaaaaa,Disabled:#ff000000,Inactive:#ff000000],Highlight:[Active:#4059aef9,Disabled:#4059aef9],HighlightedText:[Active:#ff2d577b,Disabled:#ff333333,Inactive:#fffafafa],AlternateBase:[Active:#ffe1e1e1,Disabled:#ffe1e1e1,Inactive:#ffe1e1e1],ToolTipBase:[Active:#ffffffdc,Disabled:#ffffffdc,Inactive:#ffffffdc],ToolTipText:[Active:#ff333333,Disabled:#ff333333,Inactive:#ff333333],PlaceholderText:[Active:#ff979797,Disabled:#ff979797,Inactive:#ff979797]")
qt.qpa.gtk: Returning system palette for unknown type "ToolButtonPalette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
qt.qpa.gtk: Returning system palette for unknown type "ButtonPalette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
qt.qpa.gtk: Creating palette: "CheckBoxPalette"
qt.qpa.gtk: Creating palette: "RadioButtonPalette"
qt.qpa.gtk: Returning system palette for unknown type "HeaderPalette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
qt.qpa.gtk: Returning system palette for unknown type "ItemViewPalette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
qt.qpa.gtk: Returning system palette for unknown type "MessageBoxLabelPelette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
qt.qpa.gtk: Returning system palette for unknown type "TabBarPalette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
qt.qpa.gtk: Returning system palette for unknown type "LabelPalette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
qt.qpa.gtk: Creating palette: "GroupBoxPalette"
qt.qpa.gtk: Creating palette: "MenuPalette"
qt.qpa.gtk: Returning system palette for unknown type "MenuBarPalette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
qt.qpa.gtk: Returning system palette for unknown type "TextEditPalette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
qt.qpa.gtk: Returning system palette for unknown type "TextEditPalette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
qt.qpa.gtk: Creating palette: "TextLineEditPalette"
qt.qpa.gtk: Returning system palette for unknown type "ToolTipPalette"
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
QApplication: invalid style override 'adwaita' passed, ignoring it.
	Available styles: Windows, Fusion
qt.qpa.gtk: Returning palette from cache: "SystemPalette"
QPainter::begin: Paint device returned engine == 0, type: 2
QWidget::render: Cannot render with an inactive painter
qt.gui.imageio.jpeg: Corrupt JPEG data: premature end of data segment
```

**Comment by ilya-fedin** [comment]
It says the text is #333333 while the background is #f1f1f1. That's clearly a light theme, not a dark theme.

**Comment by ilya-fedin** [comment]
Looking at Qt sources, it seems to have two freedesktop dark style implementations and the behavior is totally inconsistent depending at the environment Qt application runs in. I've created https://bugreports.qt.io/browse/QTBUG-116197 but idk whether it would get any priority and whether it would considered as a bug at all by Qt developers, they seem to really like such conditional behavior, there are checks for various variables like XDG_CURRENT_DESKTOP and different codepaths even in font rendering code...

**Comment by TomiOhl** [comment]
Nice! Really hope this will get attention from Qt devs

**Comment by ilya-fedin** [comment]
I guess we can hack this in meantime with a Qt patch... It's easy to solve by removing a level of cache from Qt and making application's setting listener to signal a change to Qt rather than applying the setting directly. That would make the behavior consistent for official builds (and leave unofficial builds without any portal dark style support apparently but I guess that's better than having the support only after the setting changes) and make the official binary to lag in some environment on theme change (as it would do a synchronous request to portals whenever gtk theme changes and portals don't answer on some system leading to 30s freeze until the request times out).

**Comment by ilya-fedin** [comment]
Done, it should be now consistent but it also means elementaryOS won't have freedesktop dark style at all given that it chooses the gtk3 platformtheme explicitly. This could be fixed only either by Qt devs or by patching the platformtheme choosing algorithm further, so xdgdesktopportal has the preference over the explicit choice...

**Comment by BachoSeven** [comment]
@ilya-fedin This update actually broke the dark/light mode toggle for me (specifically 4.9.4, where the change was released).

In detail: I use `QT_QPA_PLATFORMTHEME=qt5ct` on my system, in order to have consistent GTK/Qt theming (I use a Window Manager), and previous to this release doing this toggled the `telegram-desktop` theme:

`			gsettings set org.gnome.desktop.interface color-scheme 'prefer-dark'`
(and the respective command for `prefer-dark`)

Now, no matter what I set `QT_QPA_PLATFORMTHEME` to (also tried `gtk3` and `xdgdesktopportal`), this doesn't work anymore.

Any clues?

**Comment by TomiOhl** [comment]
F
In the meantime I can confirm this works flawlessly on elementary OS (with qt platformtheme set to gtk3)

**Comment by BachoSeven** [comment]
I've played a bit more with it, and it seems that "Match the system settings" option is not there if it's set to `qt5ct`; not sure if this has always been the case, but what I've found is that even with PLATFORMTHEME set to `gtk3`, and "match system settings" on, the aforementioned commands do not toggle the theme.

However, with `gtk3` and match-system-settings, I am able to toggle the theme with the `xsettingsd` program (https://wiki.archlinux.org/title/Xsettingsd), but this is not ideal as I need `PLATFORMTHEME=qt5ct` for consistent theming in GTK/Qt apps.

**Comment by BachoSeven** [comment]
A last observation: something which is better when setting it to `gtk3` is that the "startup" theme is correct, while the method I was employing before was only working while telegram was open.

**Comment by ilya-fedin** [comment]
Since 4.8.3 tdesktop started using the [QStyleHints::colorScheme](https://doc.qt.io/qt-6/qstylehints.html#colorScheme-prop) Qt API trusting Qt it will use the appropriate API for current desktop but there remained the portal setting listener in tdesktop that was leading to inconsistency on startup vs when dark mode is changed in runtime (only qt vs both qt & portal, i.e. whether it uses qt or portal depends on what signal was emitted later) observed in this report. Since 4.9.2 tdesktop follows only QStyleHints::colorScheme to avoid such weird behavior. tdesktop also patches Qt to always prefer portal since quite a long time, since 4.9.4 the Qt patch is extended to prefer portal harder given that this wasn't working on elementaryOS.

TL;DR the official builds should always follow portal whenever it presents and fallback to GtkSettings thanks to Qt patches. Unofficial builds will follow whatever upstream Qt decides it should follow on your system. If you use qt5ct then it's up to qt5ct and apparently qt5ct doesn't implement that. If you don't like that, feel free to switch to official build or report to Qt or qt5ct.

**Comment by escape0707** [comment]
For anyone using Gnome coming here for a temporary workaround before Qt pushed better dark theme detection behavior, you can use [legacy-gtk3-theme-scheme-auto-switcher](https://extensions.gnome.org/extension/4998/legacy-gtk3-theme-scheme-auto-switcher/) to make your gtk3 theme follows your Gnome dark mode settings. This will also work for legacy apps like Geary.

**Comment by ilya-fedin** [comment]
Has anyone reported that dark theme detection doesn't work on GNOME upstream? As far as I'm aware Qt has no such report and so there couldn't be upstream work for that.

**Comment by escape0707** [comment]
Humm, not working for me on my Arch + GNOME + Arch repo tdesktop set up. Official tdesktop binary works (of course).
